add a search field for the tag buttons in Preview

egolus opened the pull request:
this adds a search field to the top of the right hand side in Preview.
If you type into this field, the buttons get filtered.

At the moment the layout is jumping and the tests are missing

egolus committed:
add a search field for the tag buttons in Preview

egolus committed:
fixed jumping layout

egolus committed:
restore width for the scrollarea

egolus commented:
now the layout should look the same as before

egolus committed:
tests for Preview.searchField

egolus committed:
compare caseinsensitive

egolus committed:
fuzzy search for search field

egolus committed:
return key in searchField hits first visible button

if you filter the tag buttons you probably want to hit the one you find

baudren commented:
Nice work - it works well.

Two small remarks:
- It could be nice to have a greyed out text saying `filter tags`, or `search tags`, to indicate to people the use of this field. It would disappear once any text is written.
- Could also be nice to have, on the right of this text field, a small [cross](http://fortawesome.github.io/Font-Awesome/icon/times/), to remove all selected tags.
- On the topic of pressing `ctrl-F` for focusing on this field, I don't know. I did not even succeed in replacing the `Alt-something` behaviour of PySyde (for instance, we currently have to press `alt-s` to save the text, which is not optimal). I am interested in any solution you could find :)

egolus commented:
yes, these two things I wanted to do, too.

I tried with `Alt-K`, too (`Alt-F` is allready in use) but didn't really succeed

egolus committed:
add a placeholder text for the search field

this text is visible if the field is empty and doesn't have the focus.
It gives an indication what the field is for

egolus commented:
I would really like a clear button (`x`) on the search field. In Qt5 this exists out of the box. I'll look if I can implement it, but I don't have any idea how to put a widget inside of a `QLineEdit`. Do you know how to do that?

egolus committed:
new widget LineEditWithClearButton

a standard QLineEdit with an clear button added

egolus committed:
tests for LineEditWithClearButton

baudren commented:
Nice implementation - the fact that it is just a `x` and not an icon is really not an issue. I like it, and I think that this is ready to merge. Thanks for the work, @egolus

baudren committed:
remove import test line
